pkg/statsd/statsd.go

```diff
@@ -118,14 +119,32 @@ func (s *Server) RunWithCustomSocket(ctx context.Context, sf SocketFactory) erro
 	})
 
 	// 4. Start the heartbeat
-
 	if s.HeartbeatInterval != 0 {
 		hb := stats.NewHeartBeater(statser, "heartbeat", s.HeartbeatTags, s.HeartbeatInterval)
 		stage = stgr.NextStage()
 		stage.StartWithContext(hb.Run)
 	}
 
-	// 5. Start the Receiver
+	// 5. Start the Parser
+	// Open receiver <-> parser chan
+	datagrams := make(chan *Datagram, s.MaxReaders*s.ReceiveBatchSize)
```

Comment:
Do we need to make this buffered?

Reply:
We saw some inconsistency in the rate that datagrams were being read - my idea here was to buffer that out. Also gives us insight into which component is backing up. Whats your concern with making it buffered, memory usage?

Reply:
Datagrams are buffered in network stack already, I don't see why additional buffering inside of the application is needed.
CPU puts this data into its cache (read from network stack), then discards it from the cache (while it sits idle in the buffer), then puts it back again (when data is read from the buffer). Less buffering is better because there is less chance for cache thrashing. Also it is easier to understand unbuffered channels and chance of having a bug hidden by buffering (a deadlock typically) is much smaller.

Reply:
Note that we already have buffering - channels in from of aggregators are buffered. Don't need an extra layer between receiver and parser. If parsers cannot keep up then there is no point in reading more from network.

Reply:
And network stack already provides buffers to handle spikes in traffic.

Reply:
The network buffer fills in about 2 seconds, which aggregation exceeds.  The aggregator buffers are unevenly distributed.  This gives us a place to both control and monitor the amount we buffer.

Reply:
Nothing stops us from bumping up the size of the network buffer? And it is probably more memory efficient. And it is evenly distributed. And doesn't have problems I described above. And can and should be monitored. So?

Reply:
I can certainly see your argument for letting the network stack handling buffering. I was going to argue that we're having issues exceeding the network buffers... but I think you've hit the nail on the head with 

>  If parsers cannot keep up then there is no point in reading more from network.

I'd prefer to try tweaking network buffers further before we start tweaking here, so I'll remove this buffering for now.